src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -206,6 +206,8 @@ internal override CommonAnonymousTypeManager CommonAnonymousTypeManager
         /// </summary>
         internal bool IsPeVerifyCompatEnabled => LanguageVersion < LanguageVersion.CSharp7_2 || Feature("peverify-compat") != null;
 
+        internal bool IsStaticMethodGroupDelegateCacheEnabled => Feature("DisableStaticMethodGroupDelegateCache") is null;
```

Comment:
>IsStaticMethodGroupDelegateCacheEnabled [](http://example.com/codeflow?start=22&length=39)

I think we will be comfortable tying the caching to the next language version. I.e., if a user uses the next language version or higher, the caching is performed. If the language version is lower, the caching is not performed. In order to do that, we need to add a feature id into the ```MessageID``` enum, etc. For now, let's tie it to ```LanguageVersion.Preview``` version. We'll update that once the next language version is added. We actually do not need to report any errors for this feature, the language version will be used to determine if the feature is enabled or not. Once you make the switch, please make sure adding a test that caching is not performed for ```LanguageVersion.CSharp10```. And also we should have a test confirming that caching is performed for the language version returned by ```MessageIDExtensions.RequiredVersion``` API for the new feature id. The remaining tests focused on scenarios when feature is enabled can all target ```LanguageVersion.Preview```.

Reply:
I actually experimented using that approach in my previous draft PR, see https://github.com/dotnet/roslyn/pull/57882/commits/7022152dcfbf26a8c99ba12293638c279d0c0df9, and failed some [ExpressionCompiler tests](https://github.com/dotnet/roslyn/pull/58288/files#diff-ba3726047b8e688af9780ae63e60efdbd34259af1b3223429d6b049be1ede090R4184) locally (later found them not being run in CI). The reason that tests failing was that despite the `LanguageVersion` passed to the compilation being `LanguageVersion.Preview`, the compilation [ended up created](https://github.com/dotnet/roslyn/blob/e01c2352135e06927be8bf3bcbd737251e991729/src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs#L180) for the method `EvaluationContext` defaults back to `CSharp10`. Another odd feelings using that approach is the check was in `factory.Syntax.IsFeatureEnabled`, I don't think any syntax changes is introduced in this feature.

Reply:
> The reason those tests were failing was that despite the LanguageVersion passed to the compilation being LanguageVersion.Preview, the compilation ended up created for the method EvaluationContext defaults back to CSharp10.

That feels like just a test infrastructure issue that should be fixed instead of affecting the design of the feature. 

> Another odd feelings using that approach is the check was in factory.Syntax.IsFeatureEnabled, I don't think any syntax changes are introduced in this feature.

Not all features come with a new syntax. The setting is discovered from parse options, that is why the API is placed in syntax related area of the codebase. I think, we definitely want the language version to affect this feature. So, having yet another way to disable it doesn't feel particularly useful.  

If that would make you more comfortable, there is a ```CSharpCompilation.LanguageVersion```API that can be used instead.

Reply:
> That feels like just a test infrastructure issue that should be fixed instead of affecting the design of the feature.

Feel the same here. Did a Shift+F12 on `LanguangeVersion.Preview`, all results point to parseOptions.

Reply:
Filed issue #58449

Reply:
New checking using `MessageID`  please see af2f823